File: typescript/src/data-structures/queue.ts
export abstract class Queue<T> {
  abstract peek(): T;

  abstract enqueue(value: T): void;

  abstract dequeue(): T;

  abstract size(): number;

  abstract clear(): void;

  isEmpty() {
    return this.size() === 0;
  }
}

export class ArrayQueue<T> extends Queue<T> {
  protected elems: T[];

  constructor() {
    super();
    this.elems = [];
  }

  peek() {
    if (this.elems.length === 0) {
      throw new Error("Queue is empty");
    }
    return this.elems[0];
  }

  enqueue(elem: T) {
    this.elems.push(elem);
  }

  dequeue() {
    if (this.elems.length === 0) {
      throw new Error("Queue is empty");
    }
    return this.elems.shift();
  }

  size() {
    return this.elems.length;
  }

  clear() {
    this.elems = [];
  }
}

export interface QueueNode<T> {
  value: T;
  next: QueueNode<T>;
}

export class LinkedQueue<T> extends Queue<T> {
  protected head: QueueNode<T>;
  private _size: number;

  constructor() {
    super();
    this.head = undefined;
    this._size = 0;
  }

  peek() {
    if (!this.head) {
      throw new Error("Queue is empty");
    }
    return this.head.value;
  }

  enqueue(value: T) {
    const node: QueueNode<T> = {
      value,
      next: undefined
    };

    if (!this.head) {
      this.head = node;
      this._size++;
      return;
    }

    let currentNode = this.head;
    while (currentNode.next) {
      currentNode = currentNode.next;
    }
    currentNode.next = node;
    this._size++;
  }

  dequeue() {
    if (!this.head) {
      throw new Error("Queue is empty");
    }

    const { value } = this.head;
    this.head = this.head.next;
    this._size--;

    return value;
  }

  size() {
    return this._size;
  }

  clear() {
    this.head = undefined;
    this._size = 0;
  }
}

File: ruby/src/data-structures/queue.rb
# frozen_string_literal: true

EMPTY_QUEUE_EXCEPTION = 'Queue is empty'

module DataStructures
  class ArrayQueue
    def initialize
      @elems = []
    end

    def peek
      return raise EMPTY_QUEUE_EXCEPTION if @elems.length.zero?

      @elems[0]
    end

    def enqueue(value)
      @elems.push(value)
    end

    def dequeue
      return raise EMPTY_QUEUE_EXCEPTION if @elems.length.zero?

      @elems.shift
    end

    def size
      @elems.length
    end

    def clear
      @elems = []
    end

    def to_array
      @elems
    end

    def empty?
      @elems.length.zero?
    end
  end

  class LinkedQueue
    attr_reader :size

    class QueueNode
      attr_accessor :value, :next

      def initialize(value, next_node)
        @value = value
        @next = next_node
      end
    end

    def initialize
      @head = nil
      @size = 0
    end

    def peek
      return raise EMPTY_QUEUE_EXCEPTION if @head.nil?

      @head.value
    end

    def enqueue(value)
      node = QueueNode.new(value, nil)
      if @head.nil?
        @head = node
        @size += 1
        return
      end

      current = @head
      current = current.next until current.next.nil?
      current.next = node
      @size += 1
    end

    def dequeue
      return raise EMPTY_QUEUE_EXCEPTION if @head.nil?

      value = @head.value
      @head = @head.next
      @size -= 1

      value
    end

    def clear
      @head = nil
      @size = 0
    end

    def to_array
      result = []
      current = @head
      until current.nil?
        result.push(current.value)
        current = current.next
      end
      result
    end

    def empty?
      @head.nil?
    end
  end
end
